### ISSUE TITLE
consider adopting cancellation for when .provide() skips steps

### ISSUE THREAD

**Issue by dherman** [whatwg/loader#89]
When `.provide()` gives a result for a later step than the pipeline is currently at, we fulfill the earlier promise with `undefined`. This seems like a good candidate for cancellation, assuming JS promise cancellation gets worked out in time for us to adopt it.

**Comment by matthewp** [comment]
Shouldn't the earlier promise resolive with the result of provide()?

**Comment by matthewp** [comment]
Same for install.

**Comment by dherman** [comment]
> Shouldn't the earlier promise resolive with the result of provide()?

No, because the value passed to `.provide()` is the result of a different step.

**Comment by matthewp** [comment]
I'm not sure I follow.  I thought we were talking about this scenario:

```js
var origPromise = loader.import("foo");

setTimeout(function(){
  loader.provide("foo", "translate", "export var a = 1;");
});
```

If the original import is on step "fetch" when the `provide()` is called then `origPromise` should be resolved with the result of provide.

Do you disagree here?

**Comment by dherman** [comment]
> Do you disagree here?

Oh, no, that's not what this issue is referring to. (Although the result of `origPromise` won't be the actual string "export var a = 1;" but that string will continue through the pipeline to parse, link, and instantiate the module with that source.)

This issue is referring to the promise associated with the fetch step. If we were in the middle of fetching and that promise hasn't completed, and meanwhile someone calls `.provide`, we're basically pre-empting the fetch. So we should really inform the fetch that it's not needed anymore.

Does that make more sense?

**Comment by matthewp** [comment]
That makes sense, you are pre-empting the fetch but not pre-empting the entire import() promise.  The import() should be resolved with the instantiated module provided by provide().

In the old spec this did not work correct.  I've mentioned this previous here: https://github.com/whatwg/loader/issues/48

This effectively make define/set (the old provide/install) of little use.

**Comment by dherman** [comment]
Now that we've landed PR #97 I've updated this issue description.
